enhancements/declarative-index-config.md

```diff
@@ -253,34 +264,41 @@ $ cat etcd.json
     "version": "0.9.4",
     "properties":[
         {
-            "name": "olm.package",
-            "values": {
+            "type": "olm.package.provided",
+            "value": {
                 "packageName": "etcd",
                 "version": "0.9.2-clusterwide"
             }
         },
         {
-            "name":"olm.gvk",
-            "type": "provided",
-            "values":{
-                "group": "etcd.database.coreos.com", 
-                "kind": "EtcdBackup", 
+            "type": "olm.package.required",
+            "value": {
+                "packageName": "test",
+                "versionRange": ">=1.2.3 <2.0.0-0"
+            }
+        },
+        {
+            "type": "olm.gvk.provided",
+            "value": {
+                "group": "etcd.database.coreos.com",
+                "kind": "EtcdBackup",
                 "version": "v1beta2"
             }
         },
         {
-            "name":"olm.gvk",
-            "type": "required",
-            "values":{
-                "group": "testapi.coreos.com", 
-                "kind": "testapi", 
+            "type": "olm.gvk.required",
+            "value": {
+                "group": "testapi.coreos.com",
+                "kind": "Testapi",
                 "version": "v1"
             }
         },
         {
-            "name": "olm.channel",
-            "value": "singlenamespace-alpha",
-            "replaces": "etcdopertor.v0.9.2"
+            "type": "olm.channel",
+            "value": {
+                "name": "singlenamespace-alpha",
+                "replaces": "etcdoperator.v0.9.2"
```

Comment:
This coupling is unexpected to me since CSVs define a replacement irrespective of any particular channel.

Reply:
Is declarative config beholden to the CSV API?

Seems like it could be nice from a UX perspective to give users the flexibility to define `replaces` upgrade graphs differently for each channel. I suppose the idea applies to other upgrade-graph-affecting fields (e.g. skips and skipRange).

Reply:
The source of truth for replaces edges today is the `channel_entries` table, which does not necessarily give you the same replacement for a given bundle depending on which channel you're dealing with.

Reply:
> Is declarative config beholden to the CSV API?

I don't necessarily think so, I'm just pointing out that this changes the semantics of channels from what we have today; i.e. for a given set of operator versions, the set of updates can differ depending on what channel you're subscribed to.

Reply:
i think this is expressing a tension between having the config reflect exactly what was in the db (where the db content was the result of reconciling the replaces/skips chains and figuring out which channels bundles ended up in as a result of that) vs having the declarative config define where things go explicitly.

It seems like part of the confusion is around explicit vs implicit channel assignment.  A bundle can be explicitly assigned to a channel, in which case it pulls other bundles into its channel when it references them as replaces+skips, or a bundle can be implicitly part of a channel, because it was referenced by another bundle in that channel.

Would it make sense for the declarative config to make this clearer?  And should the declarative config represent all channel membership, or only explicit membership?

Reply:
> where the db content was the result of reconciling the replaces/skips chains and figuring out which channels bundles ended up in as a result of that

Not exactly. Consider the non-default modes for `opm index add` -- those produce update graphs that are independent of "replaces" as specified in the CSV manifest. Consider the index `quay.io/bluddy/foo:rep-cat`. It contains an operator (package "a-operator") with three bundles: v1, v2, and v3. All three exist in channel "beta", but channel "alpha" contains only v1 and v3. Building an index with `--mode=semver` produces the replaces chain v1<-v2<-v3 for channel beta and v1<-v3 for channel alpha. In order to make an equivalent JSON representation, we either need to require one bundle blob per channel, or define replaces as (channel, replacee).

Reply:
> one bundle blob per channel

This is effectively what the registry server does today because "channel" and "replaces" are both first-class fields in the Bundle proto.

Reply:
> one bundle blob per channel

This seems less than ideal, because many of the bundle properties truly are metadata about the bundle, not the channel upgrade graph. Does this hint at the need for separating bundle metadata from channel metadata?

Maybe this is crazy, but what if we made a new first class channel schema
```json
{
   "schema": "olm.channel",
   "name": "alpha",
   "package": "etcdopeator",
   "properties": [
      {
          "type": "olm.upgradeEdge",
          "value": {
              "from": "etcdoperator.v0.6.1",
              "to": "etcdoperator.v0.9.0"
          }
      }
   ]
}
```

It begs the question about how to handle single-bundle channels, so we would need to figure that part out.

Reply:
It's definitely not ideal -- it is tech debt. I only mentioned it here to support my point that channel-dependent "replaces" edges is something that we currently support.

For this proposal, I like it the way it is written now: an olm.channel property means that the bundle is present in the named channel, and if a "replaces" key is present, that indicates its relative position in the channel's update chain. A bundle is in more than one channel if it has more than one channel property.

Reply:
> we either need to require one bundle blob per channel, 

sounds painful for anyone who wants to promote a bundle from "fast" to "fast,stable" 

> or define replaces as (channel, replacee).

or do we just have to say that a declarative config is not valid if bundleA replaces bundleB, and bundleA+B are not in the same set of channels?

Reply:
(as an aside, such a requirement that all bundles in a replaces chain belong to the same channel is likely to run into concerns from operator teams who want to promote "v3" to stable, but do not want to promote every older version in the replaces chain, to stable.  But fundamentally "the entire chain belongs to a common channel" is an OLM requirement today, even if it's not obvious to bundle authors)

Reply:
> or do we just have to say that a declarative config is not valid if bundleA replaces bundleB, and bundleA+B are not in the same set of channels?

I may be mis-intepreting, but they don't have to both be in the same set of channels. They just need to be in the same set of channels where A replaces B.

I've got a validation for this in the WIP PR for serving the declarative config: https://github.com/joelanford/operator-registry/blob/0330c1ae7b8804f9e5a16d1e6fcbef991533838f/pkg/model/model.go#L194-L198

Note that in the `model` representation, I've built a hierarchy of pkg -> channels -> bundles, so if bundleA is in channel1 and channel2, there will be:
- pkg -> channel1 -> bundleA_1
- pkg -> channel2 -> bundleA_2

each with a different `Replaces` value, based on the `olm.channel` properties listed in the bundle.

Reply:
> want to promote "v3" to stable, but do not want to promote every older version

This should be doable as this enhancement is written. To promote to stable, all you'd have to do is add property: 
```json
{
  "type": "olm.channel",
  "value": {
    "name": "stable",
    "replaces": "<current head of stable>"
}
```

You wouldn't have to bring along the old bundles, because you can explicitly state the replaces value for this bundle for the `stable` channel.

Reply:
Just to restate, so there's no ambiguity: My  comment, which started this thread:

> This coupling is unexpected to me since CSVs define a replacement irrespective of any particular channel.

is **wrong**. **It is possible** with `opm` today to generate **valid** update graphs in which a single version of an operator has different inbound edges in different channels.

Reply:
so if it's going to be valid to have multiple "replaces(version,channel)" tuples, does that mean the following will be valid for a bundle entry in the package.json?

replaces(v1,channel1)
replaces(v2,channel2)

?

because that's not something that's possible to define in the CSV today (you only get to specify a single version to replace) right?

Reply:
It will be valid to have something like this:
```json
[{
	"type": "olm.package.provided",
	"value": {
		"name": "etcd",
		"version": "2.2.16"
	}
},{
	"type": "olm.channel",
	"value": {
		"name": "fast-v2",
		"replaces": "etcd.v2.2.15"
	}
},{
	"type": "olm.channel",
	"value": {
		"name": "stable-v2",
		"replaces": "etcd.v2.1.13"
	}
}]
```
This channel scheme would basically be:
- push every minor and patch commit to `fast-v{major}`
- only promote a minor release to `stable-v{major}` when enough patches have landed
- make new channels for major releases.

Reply:
ok, so that sounds reasonable (tying replaces to channels), but how does a bundle actually declare which channels it is a part of?  (especially since it may not use replaces at all)?  And is it valid to have a bundle say "i'm a member of channels A and B and also "replace(v1,channelC)"?  Or will "ChannelC" have to appear in the list of channels the bundle declares itself to be a member of?

Reply:
For a bundle to say "I'm in channel stable-v2", it would set property:
```json
{
	"type": "olm.channel",
	"value": {
		"name": "stable-v2"
	}
}
```

If it ALSO wants to replace another bundle in stable-v2, it would set the `value.replaces` field.
```json
{
	"type": "olm.channel",
	"value": {
		"name": "stable-v2",
		"replaces": "etcd.v2.1.13"
	}
}
```